java/daikon/chicory/Instrument24.java

```diff
@@ -0,0 +1,1757 @@
+              debugInstrument.log("  %s%n", a);
+            }
+            debugInstrument.log("mgen.getSignature: %s%n", mgen.getSignature());
+            MethodTypeDesc mtd = mm.methodTypeSymbol();
+            debugInstrument.log("mtd.descriptorString: %s%n", mtd.descriptorString());
+            debugInstrument.log("mtd.displayDescriptor: %s%n", mtd.displayDescriptor());
+          }
+
+          // Create a MethodInfo that describes this method's arguments and exit line numbers
+          // (information not available via reflection) and add it to the list for this class.
+          MethodInfo curMethodInfo = create_method_info(classInfo, mgen);
+
+          if (curMethodInfo == null) { // method filtered out!
+            // We are not going to instrument this method.
+            // We need to copy it to the output class.
+            outputMethodUnchanged(classBuilder, mm, mgen);
+            continue;
+          }
+
+          shouldInclude = true; // at least one method not filtered out
+
+          method_infos.add(curMethodInfo);
+
+          int method_info_index;
+          synchronized (SharedData.methods) {
+            method_info_index = SharedData.methods.size();
+            assert curMethodInfo != null : "@AssumeAssertion(nullness): just checked above";
+            SharedData.methods.add(curMethodInfo);
+          }
+
+          // Add entry instrumentation and instrument return instructions.
+          classBuilder.withMethod(
+              mm.methodName().stringValue(),
+              mm.methodTypeSymbol(),
+              mm.flags().flagsMask(),
+              methodBuilder ->
+                  instrumentMethod(methodBuilder, mm, mgen, curMethodInfo, method_info_index));
+        }
+      }
+    } catch (Exception e) {
+      System.err.printf("Unexpected exception encountered: %s", e);
+      e.printStackTrace();
+    }
+
+    // Copy all other ClassElements to output class (unchanged).
+    for (ClassElement ce : classModel) {
+      debugInstrument.log("ClassElement: %s%n", ce);
+      switch (ce) {
+        case MethodModel mm -> {}
+          // Copy all other ClassElements to output class (unchanged).
+        default -> classBuilder.with(ce);
+      }
+    }
+
+    // Add the class and method information to runtime so it is available
+    // as enter/exit ppts are processed.
+    classInfo.set_method_infos(method_infos);
+
+    if (shouldInclude) {
+      debug_transform.log("Added trace info to class %s%n", classInfo);
+      synchronized (SharedData.new_classes) {
+        SharedData.new_classes.add(classInfo);
+      }
+      synchronized (SharedData.all_classes) {
+        SharedData.all_classes.add(classInfo);
+      }
+    } else { // not included
+      debug_transform.log("Trace info not added to class %s%n", classInfo);
+    }
+    classInfo.shouldInclude = shouldInclude;
+  }
+
+  /**
+   * Output the given method with no changes.
```

Comment:
Output to where?  
What is the distinction between `mm` and `mgen`?

Same questions for `copyMethod` and `copyCode`.